### PR TITLE
Allow human readable duration strings in DecorationConfig duration fields.

### PR DESCRIPTION
**Pull request by cjwagner**
This PR adds a new Duration type that can parse either 'integer number of nanoseconds' or 'duration string' format and serializes to the 'duration string' format. This type is used to replace the existing DecorationConfig duration fields with backwards compatibility for existing values. Now those fields can be specified in config like `"1h15m"` instead of `4500000000000` (ns units).

/cc @stevekuznetsov @sebastienvas

### CODE REVIEW

File: prow/apis/prowjobs/v1/types.go [hunk 1]
```diff
@@ -186,6 +196,33 @@ type DecorationConfig struct {
...
+	if d == nil {
+		return nil
+	}
+	reconcile := func(t *time.Duration, m *metav1.Duration) bool {
+		tSet := *t != 0
+		mSet := m.Duration != 0
+		if tSet && mSet {
+			return false
+		}
+		if tSet {
```
Comment (sebastienvas) [inline prow/apis/prowjobs/v1/types.go]: Does this mean that if both are set, t takes precedence ?

**Comment by sebastienvas** [comment]
thanks
/lgtm
/hold

**Comment by Katharine** [comment]
I think the `_readable` suffix is pretty awkward, and likely to be confusing to users.

Couldn't we just make `timeout` and `grace_period` accept either? At this point you already have to post-process the config, so you could just make them strings (or something else that can take either), try parsing a duration, then try parsing a number.

File: prow/test/data/test-config.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 plank:
   default_decoration_config:
     timeout: 7200000000000 # 2h
```
Comment (stevekuznetsov) [inline prow/test/data/test-config.yaml]: You want to change this?

File: prow/config/config_test.go [hunk 3]
```diff
@@ -323,8 +324,10 @@ periodics:
       - "test"
       - "./..."`,
 			expected: &prowapi.DecorationConfig{
-				Timeout:     2 * time.Hour,
-				GracePeriod: 15 * time.Second,
+				Timeout:             2 * time.Hour,
+				TimeoutReadable:     metav1.Duration{Duration: 2 * time.Hour},
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: Is this valid to have both set?

File: prow/test/data/test-config.yaml (hunk 2 above)
Comment (cjwagner) [reply prow/test/data/test-config.yaml]: No, I want to test that the old format parses correctly too. I validate that the `timeout` field is parsed correctly in the explicit value test case.

**Comment by cjwagner** [comment]
I got rid of the `_readable` fields and switched to using a new Duration type that can parse either 'integer number of nanoseconds' or 'duration string' format and serializes to the 'duration string' format.
The API server shouldn't care that we are changing the type of this field right? I think it is sufficient that our components parse both the existing and new formats properly.

**Comment by stevekuznetsov** [comment]
Yeah the type is irrelevant I think, don't think `etcd` cares about how to store the record. Only one way to find out ;)

**Comment by cjwagner** [comment]
Passes tests now. (My editor's go formatting disagrees with `go fmt`...) PTAL.

File: config/jobs/kubernetes-security/generated-security-jobs.yaml [hunk 4]
```diff
@@ -456,8 +456,8 @@ presubmits:
         default_repo: kubernetes
         path_strategy: legacy
       gcs_credentials_secret: service-account
-      grace_period: 15000000000
-      timeout: 2400000000000
+      grace_period: 15s
+      timeout: 40m0s
```
Comment (stevekuznetsov) [inline config/jobs/kubernetes-security/generated-security-jobs.yaml]: nit: do we need the 0s here? That's odd

**Comment by cjwagner** [comment]
> The changes LGTM but can you try installing the current CRD, creating some, then changing the CRD and seeing if any of our clients fall over on a test cluster since the failure mode there is pretty rough?

I thought that is what we discussed here https://github.com/kubernetes/test-infra/pull/12100#issuecomment-484303827?

I agree that the failure mode could be nasty if I made a mistake and if there is a better way to test this I'm eager to do so, but the [ProwJob CRD](https://github.com/kubernetes/test-infra/blob/master/prow/cluster/prowjob_customresourcedefinition.yaml) did not change at all, so I'm not sure what you are suggesting I do to test this.

File: config/jobs/kubernetes-security/generated-security-jobs.yaml (hunk 4 above)
Comment (cjwagner) [reply config/jobs/kubernetes-security/generated-security-jobs.yaml]: Unfortunately that is how duration strings are printed: https://play.golang.org/p/xyrKD-34gfP

**Comment by stevekuznetsov** [comment]
Eek sorry, poor English -- I meant to install the CRD, create some ProwJobs (creating some) and edit the ProwJobs ("changing the CRD") with the new clients to see it function. But if you want to just roll it out to prow.k8s.io and roll back in case things get weird that also SGTM

**Comment by cjwagner** [comment]
Deploying from this PR or merging this PR then deploying master should be equivalent, but deploying to prod from a PR seems like bad practice so lets merge this and deploy from master.
If something is broken reverting is cheap.
/hold cancel

**Comment by cjwagner** [comment]
This appears to be working as intended. Our currently deployed config contains duration values in both formats and everything is being serialized to the human readable strings: https://prow.k8s.io/config

**Pull request merged**